### PR TITLE
Vj/json option to delete node and client from chef server

### PR DESCRIPTION
**Pull request by Vasu1105**
1) Added deleteChefNode option
2) Handled delete chef node and client from hosted chef.

### CODE REVIEW

File: lib/chef/azure/helpers/shared.rb [hunk 1]
```diff
@@ -88,4 +89,36 @@ def report_status_to_azure (message, status_type)
...
+
+    def delete_node
+      Chef::Log.info "Inside delete node call"
+      begin
+        Chef::Config.from_file("#{bootstrap_directory}/client.rb")
+
+        unless Chef::Config[:node_name]
+          ohai = Ohai::System.new
+          ohai.all_plugins
+          Chef::Config[:node_name] = ohai[:fqdn] || ohai[:machinename] || ohai[:hostname]
```
Comment (adamedx) [inline lib/chef/azure/helpers/shared.rb]: Can't you use Chef::ConfigFetcher instead? This should get the actual config that chef-client is truly using, and you can just use Chef::Config[node_name].
Alternatively, what about using Chef::Rest? Can you use a client_key to make a rest call for the node_name?

File: spec/unit/helpers_shared_spec.rb [hunk 2]
```diff
@@ -74,4 +74,60 @@
...
     end
   end
+end
+
+describe ChefAzure::DeleteNode do
+  include ChefAzure::Shared
+  before do
+
+  end
+
```
Comment (siddheshwar-more) [inline spec/unit/helpers_shared_spec.rb]: Remove unused code block.

File: spec/unit/helpers_shared_spec.rb [hunk 3]
```diff
@@ -74,4 +74,60 @@
...
+      expect(dummy_class.delete_node).to eq(0)
+    end
+
+    it 'return exit code 1 if node name not set' do
+      allow(dummy_class).to receive(:bootstrap_directory).and_return('/etc/chef')
+      allow(Chef::Config).to receive(:from_file).and_return("")
+      expect(dummy_class.delete_node).to eq(1)
+    end
+
+    it 'set node_name as fqdn/hostname if node net is not set in configuration file' do
```
Comment (siddheshwar-more) [inline spec/unit/helpers_shared_spec.rb]: In above spec title ` if node net is` has typo Please rename `net` to `name`

I think you can give better name to above spec like- use `fqdn` as node_name while deleting node from chef-server ...

File: spec/unit/helpers_shared_spec.rb [hunk 4]
```diff
@@ -74,4 +74,60 @@
...
+  include ChefAzure::Shared
+  before do
+
+  end
+
+  let(:dummy_class) { Class.new { extend ChefAzure::DeleteNode } }
+
+  context "delete_node" do
+    it "deletes node and client from chef server" do
+      allow(dummy_class).to receive(:bootstrap_directory).and_return('/etc/chef')
```
Comment (siddheshwar-more) [inline spec/unit/helpers_shared_spec.rb]: You can move repeated code like ^ to `before(:each) {}` block.

File: lib/chef/azure/helpers/shared.rb (hunk 1 above)
Comment (adamedx) [inline lib/chef/azure/helpers/shared.rb]: Ok, so there's not a good way for us to reliably obtain the node name, and the scenario we think is not super common. In that case, I think the right thing is for us not to try to do this. Users can use knife-azure if they are interactively managing the VM -- the `--purge` option will clean up the node / client based on the user's knowledge of the node name.

We can revisit this once Chef has better support for obtaining claims directly from the identity.

**Comment by siddheshwar-more** [comment]
As per above discussion, closing PR.

**Pull request closed without merging**